### ISSUE TITLE
torch.cdist gradients are NAN for p<1 and very small differences in a given dimension (0<delta<~e-45) 

### ISSUE THREAD

**Issue by zwacke** [pytorch/pytorch#36493]
## 🐛 Bug

When using fractional norm distances between a set of feature vectors (BS x Dim) and a set of class-centers (K x Dim) via torch.cdist(feat-vecs, class-vecs, p<1) as training objective, NANs occur in gradients when the difference between a value in dim D in a feature vector and a class vector is extremely small. There seem to be mysterious underflow issues in the backward pass for fractional norms, as the NANs do not occur when the difference is exactly 0.

## To Reproduce

1. Create sufficiently many random vectors to yield edge cases:

`import torch`
`bs = 40`
`k = 10000`
`dim = 1920`
`x = torch.randn(bs,dim).cuda()`
`w = torch.randn(k,dim).cuda()`
`x.requires_grad = True`
`w.requires_grad = True`

2. Calculate cdist for p=0.5, compute gradients and find NANs. By printing the indices in the dim dimension of the vectors and the specific values you get the idea that NANs occur for (nearly) identical values in the same dimension.

`y = torch.cdist(x,w, p=.5)`
`y.mean().backward()`
`x_nans = x.grad.ne(x.grad).nonzero()`
`w_nans = w.grad.ne(w.grad).nonzero()`

`print('Number of NANs found: ', len(x_nans), len(w_nans))`
`print('NANs found in dimensions:')`
`print([i[1].item()  for i in x_nans], [i[1].item()  for i in w_nans])`
`print('NANs found at values:')`
`print([x[x_nans[i][0],x_nans[i][1]].item() for i in range(len(x_nans))])`
`print([w[w_nans[i][0],w_nans[i][1]].item() for i in range(len(w_nans))])`

Example Output:

> 4 4
[1323, 1253, 1759, 1525] [1525, 1323, 1759, 1253]
[1.2044615745544434, 0.3567602336406708, -0.607083261013031, -0.5550623536109924]
[-0.5550623536109924, 1.2044615745544434, -0.607083261013031, 0.3567602336406708]

3. To see that it does not occur for truly identical values in a dim, you can simply replace x and y by zero-vecs and no NANs occur:
`x = torch.zeros(40,1920).float()`
`w = torch.zeros(10000,1920).float()`

4.  Alternatively, by using a computationally way slower alternative, we can avoid the NANs by adding a delta to the difference between two vectors, thereby pushing the min difference over the threshold delta. For me, delta = 1e-45 never yields NANs, while delta = 1e-46 always does. This workaround is a sad one though, because it increases training duration >>2-fold :(.

`delta = 1e-45`
`y = torch.stack([torch.norm(x[b]-w+delta, p=.5, dim=1) for b in range(x.size(0))])`

## Expected behavior

Efficient training using torch.cdist. Instead got a really nasty struggle to spot the source of those NANs. 

## Environment

 - PyTorch Version: 1.4
 - OS: Ubuntu 18.04
 - How you installed PyTorch:`conda`
 - Python version: 3.7
 - CUDA version: 10.1

## Additional Question:
Quite generally, is it possible to provide the efficient pairwise subtraction of two vectors that must be intrinsic to torch.cdist as a standalone function? I suspect, the inefficiency my workaround produces is caused by the pythonic list comprehension, while cdist seems to be implemented in C. Any suggestions for speedups I could gain myself?

cc @ezyang @gchanan @zou3519 @SsnL @albanD @gqchen

**Comment by ezyang** [comment]
This seems like a nasty numerical bug, upgrading priority

**Comment by brianhhu** [comment]
I've also noticed this with p > 1 on Pytorch 1.5 (but not 1.4) when running the similarity example in torchvision. I needed to add a relatively large eps value (1e-4) in order to not get nan gradients after the first iteration: https://github.com/brianhhu/vision/tree/fix-similarity-reference-nan-loss.

**Comment by albanD** [comment]
@brianhhu  could you open a new issue with a small code sample and specifying that it is a regression between 1.4 and 1.5 please?

**Comment by brianhhu** [comment]
Ok, created a new issue here: https://github.com/pytorch/pytorch/issues/37154

**Comment by gchanan** [comment]
it's unclear if a fix for this will fix https://github.com/pytorch/pytorch/issues/37154, because this existed in 1.4 and the other didn't.

**Comment by Baranowski** [comment]
True. Since I've assigned this to myself, I will also assign the other issue to myself, until I know whether there is a single solution to both.

**Comment by Baranowski** [comment]
Here is a simpler repro:

```
import torch

x = torch.randn(1, 2).cuda()
y = x.clone()

x = x + torch.tensor([[1.0, 0.0]], device='cuda:0')

x.requires_grad = True
y.requires_grad = True

torch.cdist(x, y, p=.5).mean().backward()

print(x.grad)
print(y.grad)
```

```
tensor([[1., nan]], device='cuda:0')
tensor([[-1., nan]], device='cuda:0')
```

This issue happens when the two coordinates in the same dim are identical (the diff is exactly 0.0), unless the entire vectors are identical, on all dimensions.

On my setup ~e-45 is the smallest representable number and that's why delta=e-46 doesn't help - it's probably rounded down to zero.

I will add a special case for diff==0.0 once I figure out what the correct return value is. That should do.

**Comment by Baranowski** [comment]
Actually, I don't understand calculus and autograd well enough to make a decisions here. The issue here is that in `aten/src/ATen/native/cuda/DistanceKernel.cu` we have:

```python
  // Special case backward when p is less than two                                                       
  struct lt_two {                                                                                        
    static __forceinline__ __device__ scalar_t backward(const scalar_t diff, const scalar_t grad, const scalar_t dist, const scalar_t p) { return dist == 0.0 ? 0 : sign(diff) * std::pow(std::abs(diff), p - 1) * grad / std::pow(dist, p - 1); }
  };
```

When `diff == 0.0`, we get `sign(diff) == 0.0` but `pow(abs(diff), p-1) == inf` which multiplied gives `NaN`.

We had internal discussion with @pearu and @nikitaved who both agreed this should throw an exception because the function is not differentiable at zero.

**Comment by ezyang** [comment]
cc @albanD, you've been working on subgradients recently, do you know what we should do here?

**Comment by nikitaved** [comment]
The function is not differentiable at zero, yes, for `p<1`, because `||x||_p` is concave then. What we can do is to return a subgradient. How does ReLU handles zeroes? We could do something very similar.

**Comment by albanD** [comment]
When the function is not differentiable but there are subgradients, we return the minimum-norm subgradient. (this is what we do for clam/rely/p-norms).

Note that usually the general formula does not handle this and so we very often return NaN. But we accept PR to change these to the right value as defined above.
You can see the PR here: https://github.com/pytorch/pytorch/pull/37337 where we're doing it for cdist and p = 2.

**Comment by nikitaved** [comment]
@Baranowski , you could mimic the behavior of a p-Norm backward, as they are equivalent. The min norm subgradient at zero is zero.

**Comment by zwacke** [comment]
Maybe fwiw, I'll drop in my personal user-side feedback (as a professional using pytorch daily) on returning NANs or subgradients or simply zeros:

While the former might be mathematically correct, it causes a troublesome situation to emerge for the user (as can be seen by my opening of the thread in the first place). I use and understand pytorch as a framework for training neural networks, i.e. most particularly for its higher level access to the autograd engine and various handy modules. In terms of usability, all provided functionalities (for me personally) are strictly only valuable if they kinda work in this context. 

Doing optimization we always minimize some form of loss, expressed as some form of distance between a target and a functional output. If a high-level wrapper for calculating distances between all vectors in an output batch of BxD and all vectors in a target batch KxD does produce exceptions at the point where the distance is zero in a single dimension (i.e. a case that does not solely appear at the very end of the target approximation), then this function itself is not only useless in the working context, but also a trap for the user. 

Doing numerical optimization (which is a set of approximative methods), daily struggles include various forms of numerical imprecision and approximative correctness so adding tiny epsilons to what would be a zero otherwise (or returing subgradient zeros that do not change any network parameter) is not a big issue for my experience as a user, while returning NANs is. The latter kinda renders the torch.cdist function useless for me, as its value of efficient under the hood batching of cross-calculating many p-distances is not accessible anymore if I manually have to supervise zeros in an intermediate step (after differences / before distances calculations). 

I believe from a user-experience philosophical point of view it is recommendable to either only provide less functionality (maybe efficient cross-difference) and mathematical exactness or more functionality and compromises on the latter. In this case I personally would prefer the latter, combined with explicit warnings in the documentation.

Just my 2cents. Other than that, thank y'all for a great framework and keep up the good work!

**Comment by albanD** [comment]
Thanks for the feedback.
I think my answer above completely aligns with what you're saying.

The main reason we have NaNs here is historical: the general gradient formula generates them and we did not consider these special cases when implementing these functions in the first place.
We are now working on removing them and replacing them with proper values (as you can see in the PR I listed).

For the particular issue here, I would agree with you that this is an underflow and we should find the cause and fix it!

**Comment by nikitaved** [comment]
@zwacke , out of sheer curiosity, could you please tell why you use `p<1`? `||x||_p` is not even a norm then...

**Comment by zwacke** [comment]
> 
> 
> @zwacke , out of sheer curiosity, could you please tell why you use `p<1`? `||x||_p` is not even a norm then, i.e. triangle inequality fails, hence the geometry of proximity to cluster centers becomes weird (local convexity fails). Maybe I do not understand something, just got curious...

Sure. When training a neural network for classification, the class score is usually derived by some form of classifier from an embedding vector. The functional form expressed by this classifier describes the score distribution which feeds into the logsoftmax which in turn results in the probability distribution predicted by the network. 
The functional form heavily influences the relative weight of the gradient amongst samples/class scores. Lp norms with p>1 and cos-similarity put more weight on large errors, which can be problematic when dealing with noisy datasets, i.e. with a significant amount of wrong samples in a training class. I have a usecase where for example the L1-norm gives more robust results and overall seems to produce nicer statistics (more decorellated features, less sensitivity to hyper-parameter settings and thus less divergence issues). 

I have furthermore read some papers concerning the behaviour of distances in high dimensional space and there are quite some successes with generalizing norms to fractional distances in clustering and logistic regression (e.g. start with Aggarwal et al., On the Surprising Behavior of Distance Metricsin High Dimensional Space, https://bib.dbvis.de/uploadedFiles/155.pdf ). Therefore I wanted to try whether fractional distances are even more useful for my usecase. 

While your point with the triangle equality is true, I spontaneously do not see where that limitation affects my usecase (I am not a mathematician though). I am not doing any linear transformations on the vectors or amongst the vectors to which I apply the distance function. Just taking two vectors as given, use distance to proxy score, take gradients... but I'd love to hear otherwise.

**Comment by nikitaved** [comment]
@zwacke , thanks for the reference, I will definitely have a look! So, you solve a classification problem and you produce probabilistic scores, right? Do you use `|| ||_p` as a distance between distributions, or as a regularization for the weights in your network?

**Comment by zwacke** [comment]
> 
> 
> @zwacke , thanks for the reference, I will definitely have a look! So, you solve a classification problem and you produce probabilistic scores, right? Do you use `|| ||_p` as a distance between distributions, or as a regularization for the weights in your network?

I use it as a distance between an embedding vector and a class-center vector. I invert this distance to yield a score (logits). This score is then fed into the standard routine of the crossentropy loss: 1) logsoftmax to produce the (normalized) class likelihood distribution and the loss is then calculated via NLL. 

It can however also be used for weight regularization. Standard weight decay is equivalent to L2 Norm weight regularization. Im not sure though,what exact effects fractional norms would have here. If we can cope with the non-convexity issues, we get more sparse networks ( Prune large weights less strongly and quickly bring small weights to zero). Compare https://stats.stackexchange.com/questions/318322/lp-regularization-with-p-1

**Comment by kurtamohler** [comment]
I'll start working on this
